dcicutils/aws_utils.py

```diff
@@ -0,0 +1,261 @@
+""" Functions to support general aws operations
+
+    Function names should generally follow a convention to include a short string
+    to indicate the service that they are designed to operate with eg. 's3' for AWS s3.
+"""
+import boto3
+import logging
+import mimetypes
+
+from botocore.exceptions import ClientError
+# from .exceptions import ()
+from .s3_utils import s3Utils
+
+###########################
+# Config
+###########################
+logging.basicConfig()
+logger = logging.getLogger(__name__)
+
+
+def s3_bucket_head(*, bucket_name, s3=None):
+    """ Gets head info for a bucket if it exists
+
+    :param bucket_name: name of the bucket - string
+    :param s3: AWS s3 client
+    :return: dict: head response or None
+    """
+    try:
+        s3 = s3 or s3Utils().s3
+        info = s3.head_bucket(Bucket=bucket_name)
+        return info
+    except ClientError:
+        return None
+
+
+def s3_bucket_exists(*, bucket_name, s3=None):
+    """ Does a bucket exist?
+
+    :param bucket_name: name of the bucket - string
+    :param s3: AWS s3 client
+    :return: boolean - True if exists, False if not
+    """
+    return bool(s3_bucket_head(bucket_name=bucket_name, s3=s3))
+
+
+def s3_bucket_object_count(bucket_name):
+    """ Number of objects in the given s3 bucket
+
+    NB: this works with locally stored credentials - not sure
+    if something needs to be added or if it is possible to create
+    a resource from a client or provide credentials if not locally stored
+
+    :param bucket_name: name of the bucket - string
+    :return: int - number of objects in bucket
+    """
+    bucket = boto3.resource('s3').Bucket(bucket_name)
+    # get only head of objects so we can count them
+    return sum(1 for _ in bucket.objects.all())
```

Comment:
Does this work recursively or will it only count folders at top level? Thinking about how it would count things in our files/wfoutput buckets for example which all have a folder based hierarchy.

Reply:
Unless I'm missing something s3 does not store objects with a hierarchy (folders are just a convenience in console display?) so this would count all objects in the given bucket regardless of folder.  I expect if empty folders were created then these would be included in the count so maybe checking that 0 size objects are not included may be warranted?  If I am correct about this then what may indeed be useful is to support passing in a prefix and only return a count of those objects that share said prefix - for example in the dcic account in the '4dn-dcic-open-data-transfer-logs' bucket if a parameter '2022-08-01' was provided then a count of the 32 object in that folder/with that shared prefix would be returned.